src/compiler/checker.ts

```diff
@@ -13881,22 +13881,28 @@ namespace ts {
             return links.resolvedJSDocType;
         }
 
-        function getSubstitutionType(baseType: Type, substitute: Type) {
-            if (substitute.flags & TypeFlags.AnyOrUnknown || substitute === baseType) {
+        function getSubstitutionType(baseType: Type, constraint: Type) {
+            if (constraint.flags & TypeFlags.AnyOrUnknown || constraint === baseType ||

```

Comment:
> `constraint === baseType`

Is this just when someone writes

```ts
T extends T ? TrueType : FalseType
```

?

Reply:
I presume so. The check was already there.